src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/ConcurrentPipeWriter.cs

```diff
@@ -57,91 +57,81 @@ public ConcurrentPipeWriter(PipeWriter innerPipeWriter, MemoryPool<byte> pool)
+            {
+                ThrowArgumentOutOfRangeException(nameof(bytes));
             }
+
+            _tailBytesBuffered += bytes;
+            _bytesBuffered += bytes;
+            _tailMemory = _tailMemory.Slice(bytes);
+            _bufferedWritePending = false;
         }
 
         public override ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
         {
-            lock (_sync)
+            if (_currentFlushTcs != null)
             {
-                if (_currentFlushTcs != null)
-                {
-                    return new ValueTask<FlushResult>(_currentFlushTcs.Task);
-                }
+                return new ValueTask<FlushResult>(_currentFlushTcs.Task);
+            }
 
-                if (_bytesBuffered > 0)
-                {
-                    CopyAndReturnSegmentsUnsynchronized();
-                }
+            if (_bytesBuffered > 0)
+            {
+                CopyAndReturnSegmentsUnsynchronized();
+            }
 
-                var flushTask = _innerPipeWriter.FlushAsync(cancellationToken);
+            var flushTask = _innerPipeWriter.FlushAsync(cancellationToken);
 
-                if (flushTask.IsCompletedSuccessfully)
+            if (flushTask.IsCompletedSuccessfully)
+            {
+                if (_currentFlushTcs != null)
                 {
-                    if (_currentFlushTcs != null)
-                    {
-                        CompleteFlushUnsynchronized(flushTask.GetAwaiter().GetResult(), null);
-                    }
-
-                    return flushTask;
+                    CompleteFlushUnsynchronized(flushTask.GetAwaiter().GetResult(), null);
                 }
 
-                // Use a TCS instead of something resettable so it can be awaited by multiple awaiters.
-                _currentFlushTcs = new TaskCompletionSource<FlushResult>(TaskCreationOptions.RunContinuationsAsynchronously);
-                var result = new ValueTask<FlushResult>(_currentFlushTcs.Task);
-
-                // FlushAsyncAwaited clears the TCS prior to completing. Make sure to construct the ValueTask
-                // from the TCS before calling FlushAsyncAwaited in case FlushAsyncAwaited completes inline.
-                _ = FlushAsyncAwaited(flushTask, cancellationToken);
-                return result;
+                return flushTask;
             }
+
+            // Use a TCS instead of something resettable so it can be awaited by multiple awaiters.
+            _currentFlushTcs = new TaskCompletionSource<FlushResult>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var result = new ValueTask<FlushResult>(_currentFlushTcs.Task);
+
+            // FlushAsyncAwaited clears the TCS prior to completing. Make sure to construct the ValueTask
+            // from the TCS before calling FlushAsyncAwaited in case FlushAsyncAwaited completes inline.
+            _ = FlushAsyncAwaited(flushTask, cancellationToken);
```

Comment:
All calls besides FlushAsyncAwaited don't acquire the lock anymore and are assumed to acquire it outside.

The more I think about it, I think there will be issues here with the Http2FrameWriter as nothing will prevent it from calling GetMemory/Advance. I'll fix that now.